nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/PutHDFS.java

```diff
@@ -372,54 +370,56 @@ public Object run() {
-                                if (conflictResponse.equals(APPEND_RESOLUTION) && destinationExists) {
-                                    fos = hdfs.append(copyFile, bufferSize);
-                                } else {
-                                    final EnumSet<CreateFlag> cflags = EnumSet.of(CreateFlag.CREATE, CreateFlag.OVERWRITE);
-
-                                    if (shouldIgnoreLocality(context, session)) {
-                                        cflags.add(CreateFlag.IGNORE_CLIENT_LOCALITY);
-                                    }
+                    session.read(putFlowFile, in -> {
+                        OutputStream fos = null;
+                        Path createdFile = null;
+                        try {
+                            if (conflictResponse.equals(APPEND_RESOLUTION) && destinationExists) {
+                                fos = hdfs.append(copyFile, bufferSize);
+                            } else {
+                                final EnumSet<CreateFlag> cflags = EnumSet.of(CreateFlag.CREATE, CreateFlag.OVERWRITE);
 
-                                    fos = hdfs.create(actualCopyFile, FsCreateModes.applyUMask(FsPermission.getFileDefault(),
-                                            FsPermission.getUMask(hdfs.getConf())), cflags, bufferSize, replication, blockSize,
-                                            null, null);
+                                if (shouldIgnoreLocality(context, session)) {
+                                    cflags.add(CreateFlag.IGNORE_CLIENT_LOCALITY);
                                 }
 
-                                if (codec != null) {
-                                    fos = codec.createOutputStream(fos);
+                                fos = hdfs.create(actualCopyFile, FsCreateModes.applyUMask(FsPermission.getFileDefault(),
+                                        FsPermission.getUMask(hdfs.getConf())), cflags, bufferSize, replication, blockSize,
+                                        null, null);
+                            }
+
+                            if (codec != null) {
+                                fos = codec.createOutputStream(fos);
+                            }
+                            createdFile = actualCopyFile;
+                            BufferedInputStream bis = new BufferedInputStream(in);
+                            StreamUtils.copy(bis, fos);
+                            bis = null;
+                            fos.flush();
+                        } catch (IOException e) {
+                            // Catch GSSExceptions and reset the resources
+                            Optional<GSSException> causeOptional = findCause(e, GSSException.class, gsse -> GSSException.NO_CRED == gsse.getMajor());
+                            if (causeOptional.isPresent()) {
+                                getLogger().warn("Error authenticating when performing file operation, resetting HDFS resources: {} ", e.getCause().getMessage());
```

Comment:
Can this be an error level log and can we pass the whole exception as an argument?

Reply:
Yeah same thing as above, I can add the error regardless (I think) and change it to error if the user needs to be alerted immediately.